### ISSUE TITLE
Batch upload images

### ISSUE THREAD

**Issue by valllllll2000** [akvo/akvo-flow#3791]
# Context
One of our partners needs to batch upload many images at once. All the images will belong to the same forms and the rest of the data for that form would have already been imported via spreadsheets. The form contains up to 40 photo questions. The first question should be a photo of the farmer but the rest of the pictures can be randomly assigned to any photo question in order.

The survey is here: TREE PLANTING IMPACT / planting site registration and monitoring

TODO: find out if the images all go to one form or multiple forms

# Given
All the images will go to one form
One form can have up to 45 image questions most of them non mandatory
The first image should be the photo of the farmer
All the other images order does not matter: it does not matter to which form question they are assigned

Partner imports data via spreadsheet and then would batch upload all images
They will place each group of images in the corresponding folder which have as name the instance to which these images should be assigned. Then all these images will be zipped together and uploaded.

# Possible steps
Fetch formId from the surveyInstanceId
For each folder, get a list of image questions
Upload images to s3
Assign each image to its question and instance

Errors: if the surveyInstanceId does not match anything in our datastore we need to show the error to the user so they can retry

# Doubts
Where in the UI do we add that feature?
How can it be later made usable by others?

**Comment by janagombitova** [comment]
I am not sure if it should be here or as a separate issue but we also need to ensure that the photo geotags are read and stored as if the photos came from the Flow app.

**Comment by Kiarii** [comment]
Talked further with Emanuel, we thought the batch image upload API will be applicable for single image upload - which could later be added to the Inspect data tab - so users can edit / change images if that be a need. We were left with some open questions whose response might suggest whether to consider supporting the front end stuff for single image upload/edit: 

- Has any other partner besides Ecosia asked for an image edit support
- Inspite of all the media data collected via flow, has there been an overarching use or analysis of such data. From the talks with Ecosia so far, have they hinted why it is so important for them to have image data, and what sort of value analysis wise it provides for them?

As for technical constraints, Mulo will look into applicable folder size limits i.,e. if a datapoint/submission can have some 45 images and on a moderate estimate there is some 20 datapoints/submissions that makes 45*20 = 900 images. So would we have hiccups uploading such a moderately sized folder? Do we need to curtail what the user can upload at any one go, how? 

On the UI we need to consider UI how to communicate 3 states: 
- image upload (and unpacking process)
- upload (and unpacking failure); what points of failure do we consider
- How long do we expect the 900 images to take to upload and unpack - how do we communicate that
- image upload success

**Comment by Kiarii** [comment]
looking into the messages part of the workflow; on the failure part of unboxing the zip and adding images to respective instances, what things can go wrong - that we would need to communicate back to the user?

**Comment by janagombitova** [comment]
@Kiarii can you share the latest mock ups here so we can review the copy (text) before implementation? Thank you

**Comment by Kiarii** [comment]
For the upload;
![image](https://user-images.githubusercontent.com/12692444/120592528-afbbb000-c446-11eb-9503-9046dd6ab8aa.png)

For the message;
![image](https://user-images.githubusercontent.com/12692444/120592619-d7127d00-c446-11eb-9742-78139629f97f.png)
- for the messages, the thinking was that we communicate 3 variations: a total success, a total failure (for these two we dont mention instances), and a partial success and partial failure. For the partial failure we would list out the failed instance ids

**Comment by janagombitova** [comment]
@muloem So I tried testing on UAt2 just now and here is what I did: 

* Created a survey with one photo question to start with: Jana tests > Testing image upload > New form
* Wanted to create new data points by data cleaning import, but the exports do not work - issue I already flagged 
* Created 3 submissions via webforms keeping the photo question empty
* Created 3 folders with the individual instance names. two folder have 1 photo. One folder has 2 photos.
* Zipped the folder 
* Used the bulk upload image option and waited for the progress bar to reach 100% (that was at 16:51) 
* Went to messages but did not see anything 
* Received this error from Sentry: https://sentry.io/organizations/akvo-foundation/issues/2444084982/?referrer=alert_email&environment=test (not sure if related) 
* Checked Messages at 16:56 - no update 
* Checked the dataset via Inspect data tab, still no photos showing

**Comment by janagombitova** [comment]
<img width="1298" alt="Screenshot 2021-06-09 at 14 42 00" src="https://user-images.githubusercontent.com/12456965/121356106-dab76f80-c930-11eb-965e-8e50d93af96a.png">
I repeated my test case just now and it worked

**Comment by janagombitova** [comment]
A few notes to further improve the issue: 
* Implement Messages - inform once the upload is completed, inform is an upload fails 
* Simplify the text in the UI 
   * `How to bulk upload images` --> should be `How to upload one or more images to existing submissions`
   * `1. Put images into folders by respective instance id` --> `1. Put images into folders per submission`
   * `Put images corresponding to an instance into a folder and name the folder by respective instance id e.g. "703219119"` --> `Put images corresponding to one submission into a folder and name the folder using the submission's Instance ID, e.g. "703219119"`
   * `2. Zip the instance folders and upload the zip file` --> `2. Zip the folders and upload the zip file`
   * `Flow will unbox the images into respective instances.` --> `Flow will unbox the images into respective submissions.`
   * `You will be notified in the messages tab when the process is done.` --> `Check the Messages tab to see if the process is done.`

**Comment by janagombitova** [comment]
Tested also the messages works as a charm. There are still the UI changes and tracking that need to happen before we push this to all users. I would propose to have Ecosia test, say it works and then push to all users. Creating a separate issue to do so:
